### PR TITLE
Shell - Improve output when running unknown subcommand

### PR DESCRIPTION
**Pull request by HavokInspiration**
Currently, if you try to run an unknown subcommand, it just outputs the current command help, which is rather confusing and not very dev-friendly.
So in light to recent efforts made to improve guidance when a user mistake is made and improve error messages, this change aims to improve that by displaying a real error message and some directions.
For instance, if I try to run `bin/cake cache cle`, I'll get this instead of the entire output of `bin/cake cache --help` :

```sh
bin/cake cache cle
```
help-unknown-subcommand

The guess is made with the `levenshtein` method. It simply takes the one with the better score. If multiple have the same score, the first one is used.

This is just a proposition. It it is accepted, maybe this can also be done with options, arguments and even root commands.

### CODE REVIEW

File: tests/TestCase/Console/ShellTest.php [hunk 1]
```diff
@@ -161,792 +161,6 @@ public function testConstruct()
...
-
-        $this->Shell->quiet('Just a test');
-    }
-
-    /**
-     * testOut method
-     *
-     * @return void
-     */
-    public function testOut()
```
Comment (dereuromark) [inline tests/TestCase/Console/ShellTest.php]: Why are so many tests removed?

File: src/Console/Shell.php [hunk 2]
```diff
@@ -472,12 +472,64 @@ public function runCommand($argv, $autoMethod = false, $extra = [])
...
     }
 
     /**
+     * Output a message in the console stating that the command can not be found and tries to guess what the user
+     * wanted to say. Output a list of available subcommands as well.
+     *
+     * @param string $command Unknown command name trying to be dispatched.
+     * @return void
+     */
+    protected function outputUnknownCommand($command)
```
Comment (markstory) [inline src/Console/Shell.php]: This feels awkward on Shell. Shell should be a starting point for userland classes. This feature feels like something that the ConsoleOptionParser or HelpFormatter should be doing.

File: tests/TestCase/Console/ShellTest.php (hunk 1 above)
Comment (markstory) [inline tests/TestCase/Console/ShellTest.php]: I second that question.

**Review by markstory — COMMENTED** [review]
I like the idea, and the error message is far more helpful. I'm not confident this should be a method on `Shell` though.

File: tests/TestCase/Console/ShellTest.php (hunk 1 above)
Comment (HavokInspiration) [inline tests/TestCase/Console/ShellTest.php]: What on earth did I do ?!
I'll fix it.

File: src/Console/Shell.php (hunk 2 above)
Comment (HavokInspiration) [inline src/Console/Shell.php]: > Shell should be a starting point for userland classes.
Fair enough.

I can put the method in `ConsoleOptionParser` to build the message to be displayed and just have a call to `out()` in `Shell`.

Comment (markstory) [inline src/Console/Shell.php]: I think the ConsoleOptionParser has either direct or indirect access to the ConsoleIo object.

Comment (HavokInspiration) [inline src/Console/Shell.php]: I checked and it does not seem to.
The previous behavior (outputting the full help of the command) that this PR replace did this : 

```
$this->out($this->OptionParser->help($command));
```
https://github.com/cakephp/cakephp/pull/10941/files#diff-d606d81f2aa3b802daa770d43e96be2fL475

Comment (markstory) [inline src/Console/Shell.php]: Ok. If `help()` returned the unknown command text wouldn't it all still work?

Comment (HavokInspiration) [inline src/Console/Shell.php]: That would keep the API cleaner indeed. 
There would three cases in `help()` then :

- known `subcommand` : returns the help for the subcommand
- unknown `subcommand` : returns the new output this PR aims to implement
- `null` subcommand passed (first argument of `ConsoleOptionParser::help()`) : display full command help

Comment (markstory) [inline src/Console/Shell.php]: I think having the optionparser handle those 3 cases is reasonable. If the logic gets too complex we can factor out the big chunks into separate classes.

**Comment by HavokInspiration** [comment]
As discussed with @markstory, I pushed a few commits which transfers the responsibility of building the output message in case of an unknown subcommand to the `ConsoleOptionParser::help()` method.

File: src/Console/ConsoleOptionParser.php [hunk 3]
```diff
@@ -735,13 +745,63 @@ public function help($subcommand = null, $format = 'text', $width = 72)
...
+        );
+        $out[] = '';
+        if ($bestGuess !== null) {
+            $out[] = sprintf('Did you mean : `%s %s` ?', $rootCommand, $bestGuess);
+            $out[] = '';
         }
-        if ($format === 'xml') {
-            return $formatter->xml();
+        $out[] = sprintf('Available subcommands for the `%s` command are : ', $rootCommand);
+        foreach ($subcommands as $subcommand) {
```
Comment (markstory) [inline src/Console/ConsoleOptionParser.php]: Should there be a newline before the list of available commands?

File: src/Console/ConsoleOptionParser.php [hunk 4]
```diff
@@ -735,13 +745,63 @@ public function help($subcommand = null, $format = 'text', $width = 72)
...
+     * @return string|null The closest name to the command submitted by the user.
+     */
+    protected function findClosestCommand($command, $subcommands)
+    {
+        $bestGuess = null;
+        foreach ($subcommands as $subcommand) {
+            $score = levenshtein($command, $subcommand);
+
+            if (!isset($bestScore) || $score < $bestScore) {
+                $bestScore = levenshtein($command, $subcommand);
```
Comment (markstory) [inline src/Console/ConsoleOptionParser.php]: Couldn't you do `$bestScore = $score`?

**Comment by HavokInspiration** [comment]
@markstory I addressed your comments. Thanks for the review.

**Comment by markstory** [comment]
Looks great. I wonder if we shouldn't do something similar for missing commands and missing options?

**Comment by HavokInspiration** [comment]
@markstory I was planning on doing the same thing for options and commands in two separate PR. Smaller chunks is easier for me to manage my time.

**Comment by HavokInspiration** [comment]
Closing. After discussing with @markstory, we decided to move this to 3.next. I'll open another PR soon.

**Comment by HavokInspiration** [comment]
Replaced in favor of #10974

**Pull request closed without merging**